Move flag to track filter context to QueryShardContext

cbuescher opened the pull request:
Currently there is a flag in the QueryParseContext that keeps track of whether an inner query sits inside a filter and should therefore produce an unscored lucene query. This is done in the parseInnerFilter...() methods that are called in the fromXContent() methods or the parse() methods we haven't refactored yet. This needs to move to the toQuery() method in the refactored builders, since the query builders themselves have no information about the parent query they might be nested in.

This PR moves the `isFilter` flag from the QueryParseContext to the recently introduces QueryShardContext. The `parseInnerFilter...` methods need to stay in the QueryParseContext for now, but they already delegate to the flag that is kept in QueryShardContext. For refactored queries (like BoolQ.B.) references to `isFilter()` are moved from the parser to the corresponding builder. Builders where the inner query was previously parsed using `parseInnerFilter...()` now use a newly introduces `toFilter(shardContext)` method that produces the nested lucene query with the filter context flag switched on.

PR is against the query refactoring branch

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -173,33 +172,29 @@ public QueryBuilder parseInnerQueryBuilder() throws IOException {
         }
     }
 
     QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOException, QueryParsingException {
```
one step further: I think we could deprecate/norelease these two parseFilter methods and make sure that our refactored queries don't use them, cause they have been moved to toFilter in the corresponding builder.

cbuescher replied:
I tried removing them already, the problem is that even though we might have already refactored a query (like AndQueryBuilder), we still need to use `parseInnerFilterToQueryBuilder()` for now because the inner queries might not have been refactored yet and still do their check of the filter flag in parse(). We might be able to replace `parseInnerFilterToQueryBuilder()` with `parseInnterQueryBuilder` after FuzzyQuery and TermsQuery are refactored, but I'm not sure.

javanna replied:
very good point, makes sense to me. Let's mark them norelease then?

cbuescher replied:
okay.

javanna commented:
I left a small comment, this looks great though! Thanks!

cbuescher commented:
@javanna Thanks, rebased and added the two //norelease comments and a meaningful log message. Maybe @jpountz wants to have a look since this concerns the filter/query merge an I might have missed something?

jpountz commented:
It looks good to me too

javanna commented:
LGTM too thanks @cbuescher
